functional/vfio-ap/run.sh

```diff
@@ -0,0 +1,168 @@
+
+trap cleanup EXIT
+
+# Check if the given function exists.
+function function_exists() {
+    [[ "$(type -t $1)" == "function" ]]
+}
+
+if ! function_exists get_test_version; then
+    source "${script_path}/../../.ci/lib.sh"
+fi
+image_version=$(get_test_version "docker_images.registry_ibm.version")
+registry_image=$(get_test_version "docker_images.registry_ibm.registry_url"):"${image_version}"
+
+cleanup() {
+    # Clean up container images
+    sudo ctr image rm $(sudo ctr image list -q) || :
+    ${container_engine} rmi -f ${test_image_name} > /dev/null 2>&1
+
+    # Destroy mediated devices
+    IFS=$'\n' read -r -d '' -a arr_dev < <( ls -1 /sys/bus/mdev/devices && printf '\0' )
+    for item in ${arr_dev[@]}; do
+        if [[ ${item//-/} =~ ^[[:xdigit:]]{32}$ ]]; then
+            echo 1 | sudo tee /sys/bus/mdev/devices/${item}/remove > /dev/null
+        fi
+    done
+
+    # Release devices from vfio-ap
+    echo 0x$(printf -- 'f%.0s' {1..64}) | sudo tee /sys/bus/ap/apmask > /dev/null
+    echo 0x$(printf -- 'f%.0s' {1..64}) | sudo tee /sys/bus/ap/aqmask > /dev/null
+}
+
+validate_env() {
+    necessary_commands=( "${container_engine}" "ctr" "lszcrypt" )
+    for cmd in ${necessary_commands[@]}; do
+        if ! which ${cmd} > /dev/null 2>&1; then
+            echo "${cmd} not found" >&2
+            exit 1
+        fi
+    done
+
+    if ! ${container_engine} ps | grep -q "${registry_name}"; then
+        echo "Docker registry not found. Installing..."
+        ${container_engine} run -d -p ${registry_port}:5000 --restart=always --name "${registry_name}" "${registry_image}"
+        # wait for registry container
+        waitForProcess 15 3 "curl http://localhost:${registry_port}"
+    fi
+
+    sudo modprobe vfio
+    sudo modprobe vfio_ap
+}
+
+# Find a mediated device name(uuid) and a device index after the device creation
+find_new_after_command() {
+    cmd=$1
+    IFS=$'\n' read -r -d '' -a arr_before_dev < <( ls -1 ${dev_base} && printf '\0' )
+    IFS=$'\n' read -r -d '' -a arr_before_mdev < <( ls -1 ${sys_device_base} && printf '\0' )
+
+    eval "${cmd}"
+
+    IFS=$'\n' read -r -d '' -a arr_after_dev < <( ls -1 ${dev_base} && printf '\0' )
+    IFS=$'\n' read -r -d '' -a arr_after_mdev < <( ls -1 ${sys_device_base} && printf '\0' )
+
+    for item in ${arr_after_dev[@]}; do
+        if [[ ! " ${arr_before_dev[*]} " =~ " ${item} " ]]; then
+            dev_index=${item}
+        fi
+    done
+    for item in ${arr_after_mdev[@]}; do
+        if [[ ! " ${arr_before_mdev[*]} " =~ " ${item} " ]]; then
+            mdev_uuid=${item}
+        fi
+    done
+}
+
+build_test_image() {
+    ${container_engine} rmi -f ${test_image_name}
+    ${container_engine} build -t ${test_image_name} ${script_path}
+    ${container_engine} push ${test_image_name}
+}
+
+create_mediated_device() {
+    # a device lastly listed is chosen
+    APQN=$(lszcrypt | tail -1 | awk '{ print $1}')
+    if [[ ! $APQN =~ [0-9]{2}.[0-9]{4} ]]; then
```

Comment:
The `lszcrypt` output shows card and domain in hex. So the regex should better use: `[[:xdigit:]]{2}.[[:xdigit:]]{4}` here.

Reply:
oh, my mistake! thanks for pointing this out.